File: q24_Swap_Nodes_in_Pairs/Swap_Nodes_in_Pairs.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} head
# @return {ListNode}
def swap_pairs(head)
  return head if head.nil? || head.next.nil?
  result = ListNode.new(0)
  current_node = result
  last_node = result
  while head && head.next
    current_node.val = head.next.val
    last_node = current_node
    current_node.next = ListNode.new(0)
    current_node = current_node.next
    current_node.val = head.val
    last_node = last_node.next
    last_node = current_node
    current_node.next = ListNode.new(0)
    current_node = current_node.next
    head = head.next.next
  end
  if head
    current_node.val = head.val
    last_node = current_node
  end
  last_node.next = nil if current_node.val.zero?
  result
end

File: 0024_Swap_Nodes_in_Pairs/Swap_Nodes_in_Pairs.py
# -*- coding: utf-8 -*-

"""
Given a linked list, swap every two adjacent nodes and return its head.

You may not modify the values in the list's nodes, only nodes itself may be changed.

Example:

Given 1->2->3->4, you should return the list as 2->1->4->3.
"""


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        result = ListNode(0)
        result.next = head
        p = result
        while p.next and p.next.next:
            n1 = p.next
            n2 = n1.next
            temp = n2.next
            p.next = n2
            n2.next = n1
            n1.next = temp
            p = n1
        return result.next
